### DISCUSSION TITLE
How to get the temp path factor from a hook executing in the controller process?

### DISCUSSION THREAD

**Discussion by vitaly-krugl** [pytest-dev/pytest-xdist#988] [General]
In a hook running in the controller process, I need to obtain the `pytest.TempPathFactory` instance that would produce the same temporary path as the fixture `tmp_path_factory` in the Worker processes of the same test run.

What I need to accomplish is to be able to save some data to a file in the worker processes in `pytest_terminal_summary` and read that data in `pytest_terminal_summary` that is executing in the controller process. I need a way for them to agree on a common directory where to store and find those files.

`pytest.TempPathFactory` sounds ideal for this, except the controller process doesn't execute fixtures, so I am hoping you can suggest another way to get the same from a hook in a controller

**Comment by vitaly-krugl** [comment]
Sorry, it's been a while. Does anyone know the answer?

**Reply by nicoddemus** [reply]
Hi,

It is not really possible today to access fixtures from hooks in a general manner, because fixtures depend on scope and might or might not be available at certain times. However if you have access to `pytest.Config` in a hook and don't mind accessing internals, you can get the `_tmp_path_factory` object: https://github.com/pytest-dev/pytest/blob/bd58c09500cdb8d36d222ce9db14b8a49b579bab/src/_pytest/tmpdir.py#L220


---

I had a similar requirement, but I solved it slightly different (pseudo-code):

```python
def pytest_configure(config):
    if not xdist.is_xdist_worker():
        config.session_tmp_dir = create_my_temporary_directory_for_session()
    else:
        config.session_tmp_dir = config.workerinput['session-tmp-dir']

def pytest_configure_node(node):    
    # This executes on the master node only: we pass on the session_tmp_dir
    # initialized during `pytest_configure`.
    node.workerinput['session-tmp-dir'] = str(node.config.session_tmp_dir)
```

With this, you have `config.session_tmp_dir` object with the unique directory, shared with workers. How you want to create the temporary directory will depend on your environment.

**Reply by vitaly-krugl** [reply]
@nicoddemus - thank you for following up. Your pseudocode is accessing `request` (`request.config.workerinput['session-tmp-dir']`) . How did you get `request` from your `pytest_configure` hookimp?

**Reply by RonnyPfannschmidt** [reply]
That looks like a mistake, you ought to use config, request is only available in a fixture

**Reply by nicoddemus** [reply]
Oops yes, thanks @RonnyPfannschmidt.

Sorry about that, updated the code.
